From 6ff7c6fc5049dcf9faa887959fcc50e08e459e17 Mon Sep 17 00:00:00 2001
From: Connor McAdams <conmanx360@gmail.com>
Date: Fri, 4 Mar 2022 22:39:06 -0500
Subject: [PATCH 04/10] ALSA: hda/ca0132 - Add ca0132_mmio_init data for
 SoundBlaster AE-9.

Modify the AE-5 ca0132_mmio_init function to add AE-9 specific writes.

Signed-off-by: Connor McAdams <conmanx360@gmail.com>
---
 sound/pci/hda/patch_ca0132.c | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/sound/pci/hda/patch_ca0132.c b/sound/pci/hda/patch_ca0132.c
index d1c66c14b1cb..2fd256bb0682 100644
--- a/sound/pci/hda/patch_ca0132.c
+++ b/sound/pci/hda/patch_ca0132.c
@@ -9265,7 +9265,7 @@ static const unsigned int ca0113_mmio_init_address_ae5[] = {
 	0x400, 0x42c, 0x46c, 0x4ac, 0x4ec, 0x43c, 0x47c, 0x4bc, 0x4fc, 0x408,
 	0x100, 0x410, 0x40c, 0x100, 0x100, 0x830, 0x86c, 0x800, 0x86c, 0x800,
 	0x804, 0x20c, 0x01c, 0xc0c, 0xc00, 0xc04, 0xc0c, 0xc0c, 0xc0c, 0xc0c,
-	0xc08, 0xc08, 0xc08, 0xc08, 0xc08, 0xc04, 0x01c
+	0xc08, 0xc08, 0xc08, 0xc08, 0xc08, 0xc04
 };
 
 static const unsigned int ca0113_mmio_init_data_ae5[] = {
@@ -9276,7 +9276,7 @@ static const unsigned int ca0113_mmio_init_data_ae5[] = {
 	0x00000057, 0x00800000, 0x00880680, 0x00000080, 0x00000030,
 	0x00000000, 0x00000000, 0x00000003, 0x00000003, 0x00000003,
 	0x00000001, 0x000000f1, 0x00000001, 0x000000c7, 0x000000c1,
-	0x00000080, 0x00880680
+	0x00000080
 };
 
 static void ca0132_mmio_init_sbz(struct hda_codec *codec)
@@ -9354,11 +9354,29 @@ static void ca0132_mmio_init_ae5(struct hda_codec *codec)
 			continue;
 		}
 
+		if (i == 22 && ca0132_quirk(spec) == QUIRK_AE9) {
+			writel(0x00880480, spec->mem_base + addr[i]);
+			continue;
+		}
+
 		writel(data[i], spec->mem_base + addr[i]);
 	}
 
-	if (ca0132_quirk(spec) == QUIRK_AE5)
+	switch (ca0132_quirk(spec)) {
+	case QUIRK_AE5:
+		writel(0x00880680, spec->mem_base + 0x1c);
 		writel(0x00880680, spec->mem_base + 0x1c);
+		break;
+	case QUIRK_AE7:
+		writel(0x00880680, spec->mem_base + 0x1c);
+		break;
+	case QUIRK_AE9:
+		writel(0x00000480, spec->mem_base + 0x1c);
+		writel(0x00880480, spec->mem_base + 0x1c);
+		break;
+	default:
+		break;
+	}
 }
 
 static void ca0132_mmio_init(struct hda_codec *codec)
@@ -9372,6 +9390,7 @@ static void ca0132_mmio_init(struct hda_codec *codec)
 		ca0132_mmio_init_sbz(codec);
 		break;
 	case QUIRK_AE5:
+	case QUIRK_AE9:
 		ca0132_mmio_init_ae5(codec);
 		break;
 	default:
-- 
2.43.0

